clients/da-vinci-client/src/main/java/com/linkedin/davinci/kafka/consumer/IngestionBatchProcessor.java

```diff
@@ -71,20 +72,36 @@ public IngestionBatchProcessor(
     this.version = Version.parseVersionFromKafkaTopicName(storeVersionName);
   }
 
+  // For testing
+  KeyLevelLocksManager getLockManager() {
+    return this.lockManager;
+  }
+
   /**
    * When {@link #lockManager} is not null, this function will try to lock all the keys
    * (except Control Messages) passed by the params.
    */
   public List<ReentrantLock> lockKeys(List<PubSubMessage<KafkaKey, KafkaMessageEnvelope, Long>> records) {
     if (lockManager != null) {
       List<ReentrantLock> locks = new ArrayList<>(records.size());
+
+      /**
+       * Collect all the keys and sort them to avoid deadlock.
+       */
+      List<byte[]> keys = new ArrayList<>(records.size());
+
       records.forEach(r -> {
         if (!r.getKey().isControlMessage()) {
-          ReentrantLock lock = lockManager.acquireLockByKey(ByteArrayKey.wrap(r.getKey().getKey()));
-          locks.add(lock);
-          lock.lock();
+          keys.add(r.getKey().getKey());
         }
       });
+      Collections.sort(keys, (o1, o2) -> ByteUtils.compare(o1, o2));
```

Comment:
Why does the sorting matter?

Reply:
Oh I think I understand... because of A/A, two batches with different orders of the same two keys would deadlock?

Reply:
This seems quite subtle and should be documented with a comment, IMHO...

Reply:
Sure, will add the javadoc to include the details.